core/src/metrics.cpp

```diff
@@ -0,0 +1,153 @@
+
+using cubos::core::Metric;
+using cubos::core::Metrics;
+using cubos::core::ScopeProfiler;
+using cubos::core::reflection::FieldsTrait;
+using cubos::core::reflection::Type;
+
+CUBOS_REFLECT_IMPL(Metric)
+{
+    return Type::create("Metric").with(FieldsTrait()
+                                           .withField("name", &Metric::mName)
+                                           .withField("line", &Metric::mLine)
+                                           .withField("file", &Metric::mFile)
+                                           .withField("value", &Metric::mValue));
+}
+
+namespace
+{
+    /// @brief Private type which stores the state of the metrics pool.
+    struct State
+    {
+        std::map<std::string, std::vector<Metric>> mMetrics; ///< Collection of metrics.
+        mutable std::mutex mMutex;                           ///< Mutex for thread-safe access to the metrics.
+    };
+} // namespace
+
+/// @brief Metrics pool state singleton. Guarantees it is initialized exactly once, when needed.
+/// @return Metrics pool.
+static State& state()
+{
+    static State instance{};
+    return instance;
+}
+
+void Metrics::metric(const Metric& metric)
+{
+    std::lock_guard<std::mutex> lock(state().mMutex);
+
+    state().mMetrics[metric.mName].emplace_back(metric);
+}
+
+std::size_t Metrics::size()
+{
+    std::lock_guard<std::mutex> lock(state().mMutex);
+
+    std::size_t total = 0;
+
+    for (const auto& [_, v] : state().mMetrics)
+    {
+        total += v.size();
+    }
+
+    return total;
+}
+
+std::size_t Metrics::sizeByName(const std::string& name)
+{
+    std::lock_guard<std::mutex> lock(state().mMutex);
+
+    if (!state().mMetrics.contains(name))
+    {
+        return 0;
+    }
+
+    return state().mMetrics.at(name).size();
+}
+
+bool Metrics::readMetric(const std::string& name, Metric& value, size_t& offset)
+{
+    std::lock_guard<std::mutex> lock(state().mMutex);
+
+    if (!state().mMetrics.contains(name))
+    {
+        return false;
+    }
+
+    if (offset >= state().mMetrics[name].size())
+    {
+        return false;
+    }
+
+    value = state().mMetrics[name].at(offset);
+    offset++;
+    return true;
+}
+
+bool Metrics::readValue(const std::string& name, double& value, std::size_t& offset)
+{
+    std::lock_guard<std::mutex> lock(state().mMutex);
+
+    if (!state().mMetrics.contains(name))
+    {
+        return false;
+    }
+
+    if (offset >= state().mMetrics[name].size())
+    {
+        return false;
+    }
+
+    value = state().mMetrics[name].at(offset).mValue;
+    offset++;
+    return true;
+}
+
+bool Metrics::readName(std::string& name, size_t& seenCount)
+{
+    std::lock_guard<std::mutex> lock(state().mMutex);
+
+    if (seenCount >= state().mMetrics.size())
+    {
+        return false;
+    }
+
+    // std::map does not support get by index
+    // so we manually loop it like this...
+    for (std::size_t idx = 0; const auto& n : state().mMetrics)
```

Comment:
This is a problem, as the map doesn't have a fixed order. It sorts its keys by lexicographical roder.
Thus, you could end up missing new names and getting old ones repeated when calling this function.
Instead of doing this, keep a separate `vector` for the known metric names.
Each time you add a metric in `::metric` with a name which isn't on the map, you push it to the vector.
Then, here you just access that vector.